File: ios/FarmaciasDeGuardiaEnSegovia/Services/ScheduleCacheService.swift
/*
 * Copyright (C) 2025  Bruno Follon (@bFollon)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import Foundation

/// High-performance cache service for parsed pharmacy schedules
/// Dramatically reduces app startup time by avoiding PDF re-parsing
/// Equivalent to Android ScheduleCacheService.kt
class ScheduleCacheService {
    static let shared = ScheduleCacheService()

    /// Current cache format version. Increment when cache structure changes.
    /// Version 2: Removed schedule info from Pharmacy.additionalInfo for ZBS regions
    /// Version 4: Switched from PDF-modification-date validity to server-version validity
    /// (client-offline-sync migration) - old caches don't decode into the new Pharmacy/
    /// PharmacySchedule shapes anyway, so the version bump forces them to be discarded.
    private let currentCacheVersion = 4

    private let fileManager = FileManager.default
    private let cacheDirectory: URL
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    private init() {
        // Create cache directory in app's documents directory
        let documentsURL = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first!
        cacheDirectory = documentsURL.appendingPathComponent("ScheduleCache", isDirectory: true)

        // Ensure cache directory exists
        if !fileManager.fileExists(atPath: cacheDirectory.path) {
            try? fileManager.createDirectory(at: cacheDirectory, withIntermediateDirectories: true)
            DebugConfig.debugPrint("📂 ScheduleCacheService: Created cache directory: \(cacheDirectory.path)")
        }
    }

    // MARK: - Cache Validation

    /// Check if cached schedules exist and are in the current cache format.
    /// Staleness relative to the server is handled separately, by ScheduleSyncService's
    /// manifest-version pre-check before this cache is ever read - this only guards against
    /// a format mismatch (e.g. a pre-migration cache that won't decode into current models).
    func isCacheValid(for location: DutyLocation) -> Bool {
        let cacheFile = getCacheFile(for: location)
        let metadataFile = getMetadataFile(for: location)

        guard fileManager.fileExists(atPath: cacheFile.path),
              fileManager.fileExists(atPath: metadataFile.path) else {
            return false
        }

        do {
            let metadataData = try Data(contentsOf: metadataFile)
            let metadata = try decoder.decode(CacheMetadata.self, from: metadataData)

            if metadata.cacheVersion != currentCacheVersion {
                DebugConfig.debugPrint("❌ ScheduleCacheService: Cache version mismatch for \(location.name) (expected: \(currentCacheVersion), found: \(metadata.cacheVersion))")
                return false
            }

            return true
        } catch {
            DebugConfig.debugPrint("❌ ScheduleCacheService: Error checking cache validity for \(location.name): \(error)")
            return false
        }
    }

    /// The server `version` currently on disk for a location, or `nil` if there's no valid
    /// cache. Used by ScheduleSyncService to decide whether a location needs re-fetching,
    /// without it needing to know anything about the on-disk cache format.
    func cachedServerVersion(for location: DutyLocation) -> Int? {
        guard isCacheValid(for: location) else { return nil }
        let metadataFile = getMetadataFile(for: location)
        do {
            let metadataData = try Data(contentsOf: metadataFile)
            let metadata = try decoder.decode(CacheMetadata.self, from: metadataData)
            return metadata.serverVersion
        } catch {
            return nil
        }
    }

    // MARK: - Cache Loading

    /// Load cached schedules for a location (if valid)
    func loadCachedSchedules(for location: DutyLocation) -> [PharmacySchedule]? {
        guard isCacheValid(for: location) else {
            return nil
        }

        let cacheFile = getCacheFile(for: location)

        do {
            let startTime = Date()
            let data = try Data(contentsOf: cacheFile)
            let cachedData = try decoder.decode(CachedSchedules.self, from: data)
            let loadTime = Date().timeIntervalSince(startTime) * 1000 // Convert to ms

            DebugConfig.debugPrint("⚡ ScheduleCacheService: Loaded \(cachedData.schedules.count) cached schedules for \(location.name) in \(Int(loadTime))ms")
            return cachedData.schedules

        } catch {
            DebugConfig.debugPrint("❌ ScheduleCacheService: Error loading cached schedules for \(location.name): \(error)")
            ErrorReportingService.shared.captureError(error, context: ["location": location.name, "operation": "loadCachedSchedules"])
            // If cache is corrupted, delete it
            deleteCacheFiles(for: location)
            return nil
        }
    }

    /// Get the cache timestamp for a location (when it was last saved)
    func getCacheTimestamp(for location: DutyLocation) -> TimeInterval? {
        let cacheFile = getCacheFile(for: location)

        guard fileManager.fileExists(atPath: cacheFile.path) else {
            return nil
        }

        do {
            let data = try Data(contentsOf: cacheFile)
            let cachedData = try decoder.decode(CachedSchedules.self, from: data)
            return cachedData.cacheTimestamp
        } catch {
            return nil
        }
    }

    // MARK: - Cache Saving

    /// Save synced schedules to cache, tagged with the server `version` they came from.
    func saveSchedulesToCache(for location: DutyLocation, schedules: [PharmacySchedule], version: Int) {
        do {
            let startTime = Date()

            let cachedData = CachedSchedules(
                locationId: location.id,
                locationName: location.name,
                schedules: schedules,
                cacheTimestamp: Date().timeIntervalSince1970
            )

            let cacheFile = getCacheFile(for: location)
            let data = try encoder.encode(cachedData)
            try data.write(to: cacheFile)

            let metadata = CacheMetadata(
                locationId: location.id,
                scheduleCount: schedules.count,
                cacheTimestamp: Date().timeIntervalSince1970,
                serverVersion: version,
                cacheVersion: currentCacheVersion
            )

            let metadataFile = getMetadataFile(for: location)
            let metadataData = try encoder.encode(metadata)
            try metadataData.write(to: metadataFile)

            let saveTime = Date().timeIntervalSince(startTime) * 1000 // Convert to ms
            let cacheSize = data.count / 1024 // KB

            DebugConfig.debugPrint("💾 ScheduleCacheService: Cached \(schedules.count) schedules for \(location.name) (version \(version)) in \(Int(saveTime))ms (\(cacheSize)KB)")

        } catch {
            DebugConfig.debugPrint("❌ ScheduleCacheService: Error saving schedules to cache for \(location.name): \(error)")
            ErrorReportingService.shared.captureError(error, context: ["location": location.name, "operation": "saveSchedulesToCache"])
        }
    }

    // MARK: - Cache Management

    /// Clear cache for a specific location
    func clearLocationCache(for location: DutyLocation) {
        deleteCacheFiles(for: location)
        DebugConfig.debugPrint("🗑️ ScheduleCacheService: Cleared cache for \(location.name)")
    }

    /// Clear all cached schedules
    func clearAllCache() {
        do {
            let files = try fileManager.contentsOfDirectory(at: cacheDirectory, includingPropertiesForKeys: nil)
            for file in files {
                try fileManager.removeItem(at: file)
            }
            DebugConfig.debugPrint("🗑️ ScheduleCacheService: Cleared all schedule caches")
        } catch {
            DebugConfig.debugPrint("❌ ScheduleCacheService: Error clearing all caches: \(error)")
            ErrorReportingService.shared.captureError(error, context: ["operation": "clearAllScheduleCache"])
        }
    }

    /// Get cache statistics for debugging
    func getCacheStats() -> [String: Any] {
        var stats: [String: Any] = [:]

        do {
            let files = try fileManager.contentsOfDirectory(at: cacheDirectory, includingPropertiesForKeys: [.fileSizeKey])
            let cacheFiles = files.filter { $0.pathExtension == "json" && !$0.lastPathComponent.contains(".meta") }
            let metadataFiles = files.filter { $0.lastPathComponent.contains(".meta.json") }

            stats["cacheDirectory"] = cacheDirectory.path
            stats["cacheFileCount"] = cacheFiles.count
            stats["metadataFileCount"] = metadataFiles.count
            stats["totalCacheSize"] = files.reduce(Int64(0)) { sum, fileURL in
                let attributes = try? fileManager.attributesOfItem(atPath: fileURL.path)
                let size = attributes?[.size] as? Int64 ?? 0
                return sum + size
            }

            // Per-location stats
            for cacheFile in cacheFiles {
                let locationId = cacheFile.deletingPathExtension().lastPathComponent
                do {
                    let data = try Data(contentsOf: cacheFile)
                    let cachedData = try decoder.decode(CachedSchedules.self, from: data)
                    stats["\(locationId)_scheduleCount"] = cachedData.schedules.count
                    stats["\(locationId)_cacheSize"] = data.count
                    stats["\(locationId)_cacheAge"] = Date().timeIntervalSince1970 - cachedData.cacheTimestamp
                } catch {
                    stats["\(locationId)_error"] = error.localizedDescription
                }
            }

        } catch {
            stats["error"] = error.localizedDescription
        }

        return stats
    }

    // MARK: - Bundled Day-Zero Data

    /// Loads the app-bundled schedule JSON for a location - the pre-first-sync fallback,
    /// shipped in the app bundle and never touched again after install. Disk cache (populated
    /// by a successful sync) always takes precedence over this; it's only consulted when the
    /// disk cache is empty/invalid. See Features/client-offline-sync.md.
    func loadBundledSchedules(for location: DutyLocation) -> LocationSchedule? {
        // Note: the BundledSchedules/ folder is a synchronized Xcode group, which flattens
        // its contents into the app bundle root at build time - there's no subdirectory to
        // look under at runtime, despite the source-tree layout suggesting otherwise.
        guard let url = Bundle.main.url(forResource: location.id, withExtension: "json") else {
            DebugConfig.debugPrint("📦 ScheduleCacheService: No bundled schedule found for \(location.name)")
            return nil
        }

        do {
            let data = try Data(contentsOf: url)
            let locationSchedule = try decoder.decode(LocationSchedule.self, from: data)
            DebugConfig.debugPrint("📦 ScheduleCacheService: Loaded bundled schedule for \(location.name) (\(locationSchedule.schedules.count) schedules)")
            return locationSchedule
        } catch {
            DebugConfig.debugPrint("❌ ScheduleCacheService: Error decoding bundled schedule for \(location.name): \(error)")
            ErrorReportingService.shared.captureError(error, context: ["location": location.name, "operation": "loadBundledSchedules"])
            return nil
        }
    }

    // MARK: - Private Helper Methods

    private func getCacheFile(for location: DutyLocation) -> URL {
        return cacheDirectory.appendingPathComponent("\(location.id).json")
    }

    private func getMetadataFile(for location: DutyLocation) -> URL {
        return cacheDirectory.appendingPathComponent("\(location.id).meta.json")
    }

    private func deleteCacheFiles(for location: DutyLocation) {
        do {
            try fileManager.removeItem(at: getCacheFile(for: location))
        } catch {
            // Ignore errors if files don't exist
        }
        do {
            try fileManager.removeItem(at: getMetadataFile(for: location))
        } catch {
            // Ignore errors if files don't exist
        }
    }
}

// MARK: - Data Structures

private struct CachedSchedules: Codable {
    let locationId: String
    let locationName: String
    let schedules: [PharmacySchedule]
    let cacheTimestamp: TimeInterval
}

private struct CacheMetadata: Codable {
    let locationId: String
    let scheduleCount: Int
    let cacheTimestamp: TimeInterval
    let serverVersion: Int
    let cacheVersion: Int
}

File: android/app/src/main/java/com/github/bfollon/farmaciasdeguardiaensegovia/services/ScheduleCacheService.kt
/*
 * Copyright (C) 2025  Bruno Follon (@bFollon)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.github.bfollon.farmaciasdeguardiaensegovia.services

import android.content.Context
import com.github.bfollon.farmaciasdeguardiaensegovia.data.DutyLocation
import com.github.bfollon.farmaciasdeguardiaensegovia.data.LocationSchedule
import com.github.bfollon.farmaciasdeguardiaensegovia.data.PharmacySchedule
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.File

/**
 * High-performance cache service for synced pharmacy schedules.
 * Dramatically reduces app startup time by avoiding a network round trip on every launch.
 */
class ScheduleCacheService(private val context: Context) {

    /**
     * Current cache format version. Increment when cache structure changes.
     * Version 1: Initial cache implementation
     * Version 3: Switched from PDF-modification-date validity to server-version validity, and
     * from multi-location to single-location cache entries (client-offline-sync migration) -
     * old caches don't decode into the new Pharmacy/PharmacySchedule shapes anyway, so the
     * version bump forces them to be discarded.
     */
    private val CURRENT_CACHE_VERSION = 3

    private val cacheDir = File(context.filesDir, "schedules")
    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = false
    }
    
    init {
        // Ensure cache directory exists
        if (!cacheDir.exists()) {
            cacheDir.mkdirs()
            DebugConfig.debugPrint("📂 Created schedule cache directory: ${cacheDir.absolutePath}")
        }
    }
    
    /**
     * Check if cached schedules exist and are in the current cache format.
     * Staleness relative to the server is handled separately, by ScheduleSyncService's
     * manifest-version pre-check before this cache is ever read - this only guards against
     * a format mismatch (e.g. a pre-migration cache that won't decode into current models).
     */
    fun isCacheValid(location: DutyLocation): Boolean {
        val cacheFile = getCacheFile(location)
        val metadataFile = getMetadataFile(location)

        if (!cacheFile.exists() || !metadataFile.exists()) {
            return false
        }

        try {
            val metadata = json.decodeFromString<CacheMetadata>(metadataFile.readText())

            if (metadata.cacheVersion != CURRENT_CACHE_VERSION) {
                DebugConfig.debugPrint("❌ Cache version mismatch for ${location.name} (expected: $CURRENT_CACHE_VERSION, found: ${metadata.cacheVersion})")
                return false
            }

            return true
        } catch (e: Exception) {
            DebugConfig.debugError("Error checking cache validity for ${location.name}", e)
            return false
        }
    }

    /**
     * The server `version` currently on disk for a location, or null if there's no valid cache.
     * Used by ScheduleSyncService to decide whether a location needs re-fetching, without it
     * needing to know anything about the on-disk cache format.
     */
    fun cachedServerVersion(location: DutyLocation): Int? {
        if (!isCacheValid(location)) return null
        return try {
            json.decodeFromString<CacheMetadata>(getMetadataFile(location).readText()).serverVersion
        } catch (e: Exception) {
            null
        }
    }

    /**
     * The timestamp (epoch millis) the cache for a location was last saved, or null if there's
     * no cache. Used for "last synced" UI display.
     */
    fun getCacheTimestamp(location: DutyLocation): Long? {
        val cacheFile = getCacheFile(location)
        if (!cacheFile.exists()) return null
        return try {
            json.decodeFromString<CachedSchedules>(cacheFile.readText()).cacheTimestamp
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Load cached schedules for a location (if valid)
     */
    fun loadCachedSchedules(location: DutyLocation): List<PharmacySchedule>? {
        if (!isCacheValid(location)) {
            return null
        }

        val cacheFile = getCacheFile(location)

        try {
            val startTime = System.currentTimeMillis()
            val cachedData = json.decodeFromString<CachedSchedules>(cacheFile.readText())
            val loadTime = System.currentTimeMillis() - startTime

            DebugConfig.debugPrint("⚡ Loaded ${cachedData.schedules.size} cached schedules for ${location.name} in ${loadTime}ms")
            return cachedData.schedules

        } catch (e: Exception) {
            DebugConfig.debugError("Error loading cached schedules for ${location.name}", e)
            ErrorReportingService.captureError(e, mapOf("location" to location.name, "operation" to "loadCachedSchedules"))
            // If cache is corrupted, delete it
            deleteCacheFiles(location)
            return null
        }
    }

    /**
     * Save synced schedules to cache, tagged with the server `version` they came from.
     */
    fun saveSchedulesToCache(location: DutyLocation, schedules: List<PharmacySchedule>, version: Int) {
        try {
            val startTime = System.currentTimeMillis()

            val cachedData = CachedSchedules(
                locationId = location.id,
                locationName = location.name,
                schedules = schedules,
                cacheTimestamp = System.currentTimeMillis()
            )

            val cacheFile = getCacheFile(location)
            cacheFile.writeText(json.encodeToString(cachedData))

            val metadata = CacheMetadata(
                locationId = location.id,
                scheduleCount = schedules.size,
                cacheTimestamp = System.currentTimeMillis(),
                serverVersion = version,
                cacheVersion = CURRENT_CACHE_VERSION
            )

            val metadataFile = getMetadataFile(location)
            metadataFile.writeText(json.encodeToString(metadata))

            val saveTime = System.currentTimeMillis() - startTime
            val cacheSize = cacheFile.length() / 1024 // KB

            DebugConfig.debugPrint("💾 Cached ${schedules.size} schedules for ${location.name} (version $version) in ${saveTime}ms (${cacheSize}KB)")

        } catch (e: Exception) {
            DebugConfig.debugError("Error saving schedules to cache for ${location.name}", e)
            ErrorReportingService.captureError(e, mapOf("location" to location.name, "operation" to "saveSchedulesToCache"))
        }
    }

    /**
     * Loads the app-bundled schedule JSON for a location - the pre-first-sync fallback,
     * shipped in the app assets and never touched again after install. Disk cache (populated
     * by a successful sync) always takes precedence over this; it's only consulted when the
     * disk cache is empty/invalid. See Features/client-offline-sync.md.
     */
    fun loadBundledSchedules(location: DutyLocation): LocationSchedule? {
        return try {
            context.assets.open("bundled_schedules/${location.id}.json").use { stream ->
                val locationSchedule = json.decodeFromString<LocationSchedule>(stream.reader().readText())
                DebugConfig.debugPrint("📦 Loaded bundled schedule for ${location.name} (${locationSchedule.schedules.size} schedules)")
                locationSchedule
            }
        } catch (e: java.io.FileNotFoundException) {
            DebugConfig.debugPrint("📦 No bundled schedule found for ${location.name}")
            null
        } catch (e: Exception) {
            DebugConfig.debugError("Error decoding bundled schedule for ${location.name}", e)
            ErrorReportingService.captureError(e, mapOf("location" to location.name, "operation" to "loadBundledSchedules"))
            null
        }
    }
    
    /**
     * Clear cache for a specific region
     */
    fun clearRegionCache(location: DutyLocation) {
        deleteCacheFiles(location)
        DebugConfig.debugPrint("🗑️ Cleared cache for ${location.name}")
    }
    
    /**
     * Clear all cached schedules
     */
    fun clearAllCache() {
        try {
            cacheDir.listFiles()?.forEach { it.delete() }
            DebugConfig.debugPrint("🗑️ Cleared all schedule caches")
        } catch (e: Exception) {
            DebugConfig.debugError("Error clearing all caches", e)
        }
    }
    
    /**
     * Get cache statistics for debugging
     */
    fun getCacheStats(): Map<String, Any> {
        val stats = mutableMapOf<String, Any>()
        
        try {
            val files = cacheDir.listFiles() ?: emptyArray()
            val cacheFiles = files.filter { it.extension == "json" && !it.name.endsWith(".meta.json") }
            val metadataFiles = files.filter { it.name.endsWith(".meta.json") }
            
            stats["cacheDirectory"] = cacheDir.absolutePath
            stats["cacheFileCount"] = cacheFiles.size
            stats["metadataFileCount"] = metadataFiles.size
            stats["totalCacheSize"] = files.sumOf { it.length() }
            
            // Per-region stats
            cacheFiles.forEach { cacheFile ->
                val regionId = cacheFile.nameWithoutExtension
                try {
                    val cachedData = json.decodeFromString<CachedSchedules>(cacheFile.readText())
                    stats["${regionId}_scheduleCount"] = cachedData.schedules.size
                    stats["${regionId}_cacheSize"] = cacheFile.length()
                    stats["${regionId}_cacheAge"] = System.currentTimeMillis() - cachedData.cacheTimestamp
                } catch (e: Exception) {
                    stats["${regionId}_error"] = e.message ?: "Unknown error"
                }
            }
            
        } catch (e: Exception) {
            stats["error"] = e.message ?: "Unknown error"
        }
        
        return stats
    }
    
    // Private helper methods
    
    private fun getCacheFile(location: DutyLocation): File {
        return File(cacheDir, "${location.id}.json")
    }
    
    private fun getMetadataFile(location: DutyLocation): File {
        return File(cacheDir, "${location.id}.meta.json")
    }
    
    private fun deleteCacheFiles(location: DutyLocation) {
        try {
            getCacheFile(location).delete()
            getMetadataFile(location).delete()
        } catch (e: Exception) {
            DebugConfig.debugError("Error deleting cache files for ${location.name}", e)
        }
    }
    
    // Data classes for serialization
    
    @Serializable
    private data class CachedSchedules(
        val locationId: String,
        val locationName: String,
        val schedules: List<PharmacySchedule>,
        val cacheTimestamp: Long
    )

    @Serializable
    private data class CacheMetadata(
        val locationId: String,
        val scheduleCount: Int,
        val cacheTimestamp: Long,
        val serverVersion: Int,
        val cacheVersion: Int
    )
}
